### PR TITLE
Fix date time picker

### PR DESCRIPTION
**Pull request by rafbgarcia**

### CODE REVIEW

File: components/DatePicker/DatePicker.jsx [hunk 1]
```diff
@@ -47,37 +47,48 @@ export default class DatePicker extends React.Component<Props> {
   }
 
   state = {
-    open: false,
     valid: true,
   }
 
+  props: Props
```
Comment (drborges) [inline components/DatePicker/DatePicker.jsx]: You can drop this props definition if you want. The class already `extends React.Component<Props>`, defining an empty `props: Props` is redundant in this case.

File: components/DatePicker/DatePicker.jsx [hunk 2]
```diff
@@ -47,37 +47,46 @@ export default class DatePicker extends React.Component<Props> {
...
 
   state = {
-    open: false,
     valid: true,
   }
 
   componentWillMount() {
-    const defaultValue = moment(this.props.defaultValue || "")
-    if(defaultValue.isValid()) this.handleOnChange(defaultValue)
+    this.props.onChange(this.defaultValue())
```
Comment (drborges) [inline components/DatePicker/DatePicker.jsx]: why do we need to trigger an `onChange` when the component is about to be mounted?

File: components/DatePicker/DatePicker.jsx [hunk 3]
```diff
@@ -47,37 +47,46 @@ export default class DatePicker extends React.Component<Props> {
...
+    const { timeFormat, dateFormat } = this.props
+    return timeFormat && timeFormat !== "" ? "YYYY-MM-DDTHH:mm:00Z" : dateFormat
   }
 
   handleOnChange = (dateTime) => {
-    const {timeFormat} = this.props
-    if(dateTime.format && dateTime.isValid()) {
-      this.setState({valid: true})
-      const dt = this.props.timeZone ? moment.tz(dateTime) : dateTime
-      setTimeout(() => {
```
Comment (drborges) [inline components/DatePicker/DatePicker.jsx]: I'm not really sure why there's a `setTimeout` in here, at a first glance it seems unnecessary, but @thatjsguy would know better. If that's somehow a workaround the fact the `setState` call is async, we should instead be using the `setState` onComplete callback, as in:

```jsx
const dt = this.props.timeZone ? moment.tz(dateTime) : dateTime
const onComplete = () => {
  this.props.onChange(dt.format(timeFormat && timeFormat !== "" ? "YYYY-MM-DDTHH:mm:00Z" : "YYYY-MM-DD"))
}

this.setState({ valid: true }, onComplete)
```

File: components/DatePicker/DatePicker.jsx [hunk 4]
```diff
@@ -91,51 +100,58 @@ export default class DatePicker extends React.Component<Props> {
...
         </span>
-      </span>
+      </div>
     )
   }
 
-  render() {
-    const {
+  dateTimeProps = () => {
+    let {
```
Comment (drborges) [inline components/DatePicker/DatePicker.jsx]: Our ESlint isn't properly configured yet, but we'd prefer `const` over `let` for constant values.

File: components/DatePicker/DatePicker.jsx [hunk 5]
```diff
@@ -91,51 +100,58 @@ export default class DatePicker extends React.Component<Props> {
...
+      ),
       closeOnSelect,
       dateFormat,
       inputProps,
       onChange: this.handleOnChange,
       renderInput: this.renderInput,
       timeFormat,
       timeZone,
+      value: this.defaultValue(),
+      ...props,
```
Comment (drborges) [inline components/DatePicker/DatePicker.jsx]: 👍

File: components/DatePicker/DatePicker.jsx [hunk 6]
```diff
@@ -91,51 +100,58 @@ export default class DatePicker extends React.Component<Props> {
...
     }
+  }
 
-    const formattedDate = moment(defaultValue || value).format(dateFormat)
-    dateTimeProps.value = formattedDate === "Invalid date" ? "" : formattedDate
+  defaultValue = () => {
```
Comment (drborges) [inline components/DatePicker/DatePicker.jsx]: 🎉

File: components/DatePicker/DatePicker.jsx (hunk 2 above)
Comment (rafbgarcia) [reply components/DatePicker/DatePicker.jsx]: If I recall the bug well, it won't trigger React Form's change event and the default value won't be applied if we don't do it.

File: components/DatePicker/DatePicker.jsx (hunk 3 above)
Comment (rafbgarcia) [reply components/DatePicker/DatePicker.jsx]: That's removed code...

Comment (thatjsguy) [reply components/DatePicker/DatePicker.jsx]: @drborges exactly what that was originally for - this is better 👍

File: components/DatePicker/DatePicker.jsx (hunk 2 above)
Comment (drborges) [reply components/DatePicker/DatePicker.jsx]: yeah, looks like this causes the credit application to break

@rafbgarcia can you share the steps to reproduce the bug this was supposed to fix? I feel like there must be an alternative, triggering user input events during mounting like this is somewhat unconventional and creates unexpected side effects.

**Pull request merged**